R/zchunk_LA1321.cement.R

```diff
@@ -52,72 +63,369 @@ module_energy_LA1321.cement_DISABLED <- function(command, ...) {
+      repeat_add_columns(tibble(year = ADDITIONAL_YEARS)) %>%
+      select(-old.year) %>%
+      bind_rows(L1321.in_EJ_R_cement_F_Y_base) ->
+      L1321.in_EJ_R_cement_F_Y
+
+    # Calculate remaining industrial energy use (input), subtracting cement production energy from energy balances
+    # ------------------------------------------------------------------------------------------------------------
+
+    # Subtract input energy to cement sector from industrial energy
+    L132.in_EJ_R_indenergy_F_Yh %>%
+      rename(ind.value = value) %>%
+      left_join(select(L1321.in_EJ_R_cement_F_Y, -sector), by = c("GCAM_region_ID", "fuel", "year")) %>%
+      mutate(value = ind.value - value) %>%
+      select(-ind.value) ->
+      L1321.in_EJ_R_indenergy_F_Yh_NAs
+
+    # Replace NA values in sectors with no match in cement with original values from L132.in_EJ_R_indenergy_F_Yh
+    L1321.in_EJ_R_indenergy_F_Yh_NAs %>%
+      filter(is.na(value)) %>%
+      select(-value) %>%
+      left_join(select(L132.in_EJ_R_indenergy_F_Yh, -sector), by = c("GCAM_region_ID", "fuel", "year")) %>%
+      bind_rows(filter(L1321.in_EJ_R_indenergy_F_Yh_NAs, !is.na(value))) ->
+      L1321.in_EJ_R_indenergy_F_Yh_negbio
+
+    # This dataset may now have negative values. If it's biomass, doesn't matter for energy balances; set the rest-of-industry to exogenous minimum value
+    L1321.in_EJ_R_indenergy_F_Yh_negbio %>%
+      filter(value < 0, fuel == "biomass") %>%
+      mutate(value = energy.MIN_IN_EJ_IND) %>%
+      bind_rows(filter(L1321.in_EJ_R_indenergy_F_Yh_negbio, value >= 0 | fuel != "biomass")) ->
+      L1321.in_EJ_R_indenergy_F_Yh
+
+    # Negative values in non-bio industrial energy use are problematic and have to be zeroed out
+    # In the below method, these negative values are zeroed in industrial energy and subtracted from positive values in fossil fuel use for cement
+    # and then offset with a positive adjustment to cement biomass fuel use. This preserves the total energy balances while removing negative values
+    # ----------------------------------------------------------------------------------------------------------------------------------------------
+
+    # Currently offending negative values are mostly developing regions with (probably incorrectly) low bio shares
+    # Check if any of these values exist and then conditionally perform the adjustments
+    if( any( L1321.in_EJ_R_indenergy_F_Yh[,"value"] < 0 ) ){
+
+      # Subset regions and years where any fuels are negative
+      # and aggregate those negative values (in case one region/year has multiple fuels) as a positive adjustment to biomass
+      L1321.in_EJ_R_indenergy_F_Yh %>%
+        filter(value < 0) %>%
+        mutate(sector = "cement") ->
+        L1321.cement_adj_neg
+
+      L1321.cement_adj_neg %>%
+        mutate(fuel = "biomass") %>%
+        group_by(GCAM_region_ID, sector, fuel, year) %>%
+        summarise(value = sum(value) * -1) %>%
```

Comment:
This seems like a bunch of work that might be better accomplished with a single pipeline that computes `abs(sum(value))`? Not sure, but consider if there's a way to streamline, eliminating the `if` above

Reply:
I don't think you can use a single pipeline here because it's adding and subtracting on only some rows and at different levels of aggregation of fuel. You might be able to use `abs(sum(value))`, `left_join` and then `mutate` but you still need to add the summed fossil fuels onto biomass beforehand using a couple `if_else` in lieu of separate frames. I spent like 20 min with it and didn't find anything more elegant (easily could be wrong though!). 

I think you're right the `if` isn't needed though, I tested the false case and it just binds empty data frames back to the original and spits it back out.